engine/src/main/scala/cromwell/engine/workflow/WorkflowDockerLookupActor.scala

```diff
@@ -0,0 +1,298 @@
+
+/**
+  * Ensures docker hash consistency throughout a workflow.
+  *
+  * Caches successful docker hash lookups and serve them to subsequent identical requests.
+  * Persists those hashes in the database to be resilient to server restarts.
+  *
+  * Failure modes:
+  * 1) Failure to load hashes from the DB upon restart.
+  * 2) Failure to parse hashes from the DB upon restart.
+  * 3) Failure to write a hash result to the DB.
+  * 4) Failure to look up a docker hash.
+  * 5) Timeout from DockerHashActor.
+  *
+  * Behavior:
+  * 1-2) Return a terminal lookup failure for all pending requests, transition to a permanently Failed state in which any
+  *      future requests will immediately return lookup failure.  The JobPreparation actor should fail in response to this
+  *      lookup termination message, which in turn should fail the workflow.
+  * 3-4) Return lookup failure for the current request and all pending requests for the same tag.  Any future requests
+  *      for this tag will be attempted again.
+  * 5)   Return lookup failure for all pending requests.  Any future requests for these tags will be attempted again.
+  */
+
+class WorkflowDockerLookupActor private[workflow](workflowId: WorkflowId, val dockerHashingActor: ActorRef, startMode: StartMode, databaseInterface: SqlDatabase)
+  extends LoggingFSM[WorkflowDockerLookupActorState, WorkflowDockerLookupActorData] with DockerClientHelper {
+
+  implicit val ec = context.system.dispatchers.lookup(Dispatcher.EngineDispatcher)
+
+  // Amount of time after which the docker request should be considered lost and sent again.
+  override protected def backpressureTimeout: FiniteDuration = 10 seconds
+  // A multiplier for the amount of time to wait when we get a Backpressure response before sending the request again.
+  // This effectively bounds the jitter.
+  override protected def backpressureRandomizerFactor: Double = 0.5D
+
+  context.become(dockerReceive orElse receive)
+
+  startWith(
+    stateName = if (startMode == RestartExistingWorkflow) AwaitingFirstRequestOnRestart else Running,
+    stateData = WorkflowDockerLookupActorData.empty
+  )
+
+  // `AwaitingFirstRequestOnRestart` is only used in restart scenarios.  This state waits until there's at least one hash
+  // request before trying to load the docker hash mappings.  This is so we'll have at least one `JobPreparationActor`
+  // reference available to message with a terminal failure in case the reading or parsing of these mappings fails.
+  when(AwaitingFirstRequestOnRestart) {
+    case Event(request: DockerHashRequest, data) =>
+      loadDockerHashStoreEntries()
+      goto(LoadingCache) using data.addHashRequest(request, sender())
+  }
+
+  // Waiting for a response from the database with the hash mapping for this workflow.
+  when(LoadingCache) {
+    case Event(DockerHashStoreLoadingSuccess(dockerHashEntries), data) =>
+      loadCacheAndHandleHashRequests(dockerHashEntries, data)
+  }
```

Comment:
Looks like if we receive hash requests in this state they won't be recorded in the data ?